### DISCUSSION TITLE
Questions about using MOOSE Materials with custom Material files

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#24358] [Q&A General] [answered]
I read through the documentation and believe I understand how the material file system works. Within an input file, you can declare a block with your material values, e.g

```
  [block_name]
    type = customMaterial
    constant1= 0.2
  []
```

and then within the custom material, you get the value from the user, i.e

```
 /// Get Parameter from user, name in input file is in quotes
_user_constant1(getParam<Real>("constant1")),
```

the variable _user_constant1 can now be used within the computeQpProperties() e.g

`auto _calc_in_method = _user_constant1*5;`

 to compute whatever you want, and then you send it back out with, say

`_calc_in_method(declareADProperty<Real>("calculated_value")),`

In this case, any kernel that wants to use the value we calculated here will do this:

`_necessary_constant(getADMaterialPropertyByName<Real>("calculated_value"))`

then, _necessary_constant can be used within the computeQpresidual method.

Do I understand how this works correctly? And in the above, is the "block_name" relevant or just for the user using the input file? I asked because of the following questions.

*********************************************************************************************************************************

Now for my first question- how can I use/get material properties defined by MOOSE already? I saw the following in someone's custom code:

```
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),

    _stress(getADMaterialPropertyByName<RankTwoTensor>(_base_name + "stress")),
    _cto(getADMaterialPropertyByName<RankFourTensor>(_base_name + "elasticity_tensor")),
    _strain(getADMaterialPropertyByName<RankTwoTensor>(_base_name + "mechanical_strain")),
```

In the input file, they have this:

```
  [elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 30000
    poissons_ratio = 0.2
  []
  
  [strain]
    type = ADComputeSmallStrain
  []
  
  [stress]
    type = ADComputeLinearElasticStress
  []
```

I want to understand how/why this code works. In the constructor, they "get" those material properties, which I assume are declared somewhere. However, those names- "stress", "elasticity_tensor", and "mechanical_strain". Where are they defined? I initially thought that "elasticity_tensor" was gotten from the name of the block, and same for "stress", but then that wouldn't make sense since "mechanical_strain" is not the name of the block in the input file, but they used that name to get the property. Is the block name relevant at all? Or how were these material values gotten?

I tried to look through the source code, but in ComputeSmallStrain.C, I do not see any `declareProperty` in the constructor. So I'm not sure how to know what name to call the material by when "getting" it.

Also, could you break down the possible use/purpose of the whole "base_name" thing? I don't understand its purpose.

**********************************************************************************************************************************

Following up on this question, if I were to want to declare a RankFourTensor tensor to use in my custom material file, how would I do so in the input file? I'm not very clear on the documentation (https://mooseframework.inl.gov/docs/doxygen/moose/classRankFourTensorTempl.html) but I'm assuming it would be something along the lines of:

```
[./material1_elasticitytensor]
  type = RankFourTensor
  input= '1.684e5 0.176e5 0.176e5 1.684e5 0.176e5 1.684e5 0.754e5 0.754e5 0.754e5'
  fill_method = symmetric9
  property =  betaphase_elasticity_tensor
[../]
```

So that in my custom material file, I can now do:

`_betaphase(getParam<RankFourTensor>("betaphase_elasticity_tensor")),`

Is this the correct syntax?

**Comment by GiudGiud** [comment, marked as answer]
Hello

For question 1, this 
    _stress(getADMaterialPropertyByName<RankTwoTensor>(_base_name + "stress")),
is retrieving a stress material property that is known to be declared with this name by the stress model.
The tensor mechanics module is using this convention for naming the stress, so materials are using that same name as well, without an explicit declaration in the input file of what the stress name will be

same for strain, elasticity tensor etc

The base_name is to disambiguate if there are multiple stress models. One can then pass different a base_name to retrieve the stress material property of interest. 

Question 2:
_betaphase(getParam<RankFourTensor>("betaphase_elasticity_tensor")),
this is not correct. This retrieves a tensor directly, not through the material property system.
You want:
_betaphase(getMaterialPropertyByName<RankTwoTensor>("betaphase_elasticity_tensor")),

and _betaphase must be declared in the header as a material property

Guillaume

**Reply by richmondodufisan** [reply]
So essentially all stress-related materials can be retrieved with "stress", and to differentiate them if I use multiple, I can add a base_name to each's material block? Is this convention explained anywhere in the documentation? Asking because I was wondering about materials in other modules, are their naming conventions also explained anywhere?

Apart from clarity for the user of the input file, is the block name relevant?

**Reply by GiudGiud** [reply]
the block name is generally not relevant, it s the name of the Material not the Material Property. 

Naming conventions will depend on the module, there's no universal rule

**Reply by richmondodufisan** [reply]
Thank you. A few more questions:

1. I believe I followed the steps properly and cannot figure out why I am getting this error:

`error: 'MaterialProperty< <template-parameter-1-1> >::MaterialProperty(const MaterialProperty< <template-parameter-1-1> >&) [with T = RankFourTensorTempl<double>]' is private within this context`

I have this in my custom material file header:

```
#pragma once

#include "DerivativeMaterialInterface.h"
#include "RankTwoTensor.h"
#include "RankFourTensor.h"
#include "Eigen/Core"
#include "Eigen/Dense"
#include <unsupported/Eigen/CXX11/Tensor>
#include "Material.h"

class Custom : public DerivativeMaterialInterface<Material>
{
public:
  static InputParameters validParams();

  CustomMaterial(const InputParameters & parameters);

protected:
  ...
  
  // just for debugging
  MaterialProperty<Real> & _seeoutput;

  // Materials calculated by this custom material
  MaterialProperty<RankTwoTensor> & _cust_eigenstr;
  MaterialProperty<RankFourTensor> & _elasticity_eta;
  
};
```

And in my source file, I have this:

```
#include "CustomMaterial.h"
#include <Eigen/Dense>
#include <cmath>

registerMooseObject("myApp", CustomMaterial);

InputParameters
CustomMaterial::validParams()
{
  ...
  
  return params;
}

CustomMaterial::CustomMaterial(const InputParameters & parameters)
  : DerivativeMaterialInterface<Material>(parameters),
	
         ...

	_seeoutput(declareProperty<Real>("debug_purposes")),
	_cust_eigenstr(declareProperty<RankTwoTensor>("e_ij_0")),
	_elasticity_eta(declareProperty<RankFourTensor>("c_ijkl"))
	
{
}
```

However, I get an error when I try to compile, saying

```
error: 'MaterialProperty< <template-parameter-1-1> >::MaterialProperty(const MaterialProperty< <template-parameter-1-1> >&) [with T = RankFourTensorTempl<double>]' is private within this context
   32 |  _elasticity_eta(declareProperty<RankFourTensor>("elasticity_eta"))
```

I am not really sure what this means. I tried to follow the format in the documentation. Should the variable/attribute "_elasticity_eta" have been declared publicly? 

Also, I got a similar error for a kernel trying to "get" these properties.


***********************************************************************************************************************************************

Secondly, I wanted to ask about the "MaterialADConverter" system. I am making non-AD materials because one of the kernels is a slightly edited version of a kernel from the phase-field module, and I derived the Jacobian and coded it there. However, I do not want to derive the Jacobian for another custom kernel I created from scratch, and so I want to convert the materials in my custom Material file from non-AD to AD.

I have this in my input file:

```
[convert_to_AD]
    type = MaterialADConverter
	ad_props_in = 'c_ijkl e_ij_0 mechanical_strain'
	ad_props_out = 'ADc_ijkl ADe_ij_0 ADmechanical_strain'
[]
```


In the AD Kernel, will I then "get" the AD version using the new name like this?

`_e_ij_0(getADMaterialPropertyByName<Real>("ADe_ij_0")),`

And for the ADmechanical_strain, would it be better to just make another material block for ADComputeSmallStrain? The "mechanical_strain" there is intended to be from ComputeSmallStrain which already exists in my input file

**Reply by GiudGiud** [reply]
For the second question, I would not recommend doing that.
The outgoing "AD" version does not contain the derivaties, it's just passing the Real as an ADReal, with 0 for all derivatives

**Reply by richmondodufisan** [reply]
Thanks! I should probably just work on the Jacobians then (or is there a way to convert from AD to non-AD so I can make my material class AD)?

Also, I figured out the first one- for posterity's sake, there was a missing "&" in the header of one of the kernels that used that material, and when I fixed that the error somehow disappeared.

**Reply by GiudGiud** [reply]
Ok I m not surprised for the first one.
This is what I was looking for in the code you pasted, but it seemed correct there, there was no copy?
MaterialPropertyValues may not be copied, you can only take a reference to them. This was accomplished by making the copy-constructor private, hence the compilation error message. 

It's very easy to convert from AD to non AD, in fast the same MaterialADConverter can do that (using ad_props_in and reg_props_out)

**Reply by richmondodufisan** [reply]
Got it, that makes sense. I think I didn't update the compiled version or something.

I'll try the MaterialADConverter, thanks.

Also, another follow up on the use of RankFourTensor & RankTwoTensor. As stated earlier, I have a block in my input file:

```
[./material1_elasticitytensor]
  type = RankFourTensor
  input= '1.684e5 0.176e5 0.176e5 1.684e5 0.176e5 1.684e5 0.754e5 0.754e5 0.754e5'
  fill_method = symmetric9
  property =  betaphase_elasticity_tensor
[../]
```

However, I get this error

```
*** ERROR ***
A 'RankFourTensor' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```

what do I need to do to be able to use this?

*************************************************************************************************************************************
Alternatively, I could use the ComputeElasticityTensor. The reason I didn't use it earlier was because I still had some confusion over the "base_name" syntax, and I have multiple elasticity tensors.

If I change the above to ComputeElasticityTensor, is this the correct syntax to send and retrieve it in the material file system?

```
[./beta_elasticity_tensor]
  type = ComputeElasticityTensor
  C_ijkl = '1022726 113636 113636 1022726 454545'
  fill_method = beta_phase
[../]

[./alpha_elasticity_tensor]
  type = ComputeElasticityTensor
  C_ijkl = '1022726 113636 113636 1022726 454545'
  fill_method = axisymmetric_rz
  base_name = alpha_phase
[../]
```

and then I can "get" them with

```
_c_beta(getADMaterialPropertyByName<RankFourTensor>("beta_phase_" + "elasticity_tensor")),
_c_alpha(getADMaterialPropertyByName<RankFourTensor>("alpha_phase_" + "elasticity_tensor")),
```

since I wrote the "base_name" directly, would I still need the below code?

`_base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),`

I realize I still don't fully understand how it works.

**Reply by GiudGiud** [reply]
I was looking but I cant recall how to input a RankFourTensor directly in the input. It s not with `type = RankFourTensor`

  `fill_method = beta_phase`
is not what you want. Maybe
```
  fill_method = axisymmetric_rz
  base_name = beta_phase
```

So you have two base_names essentially? Like you are getting the same property but for two phases in the same object. 
I think it s fine to hardcode beta/alpha_base instead of base_name

**Reply by richmondodufisan** [reply]
oh yeah that was a typo. I meant:

```
[./beta_elasticity_tensor]
  type = ComputeElasticityTensor
  fill_method = symmetric9
  C_ijkl = '1111 1122 1133 2222 2233 3333 2323 1313 1212'
  base_name = beta_phase
[../]

[./alpha_elasticity_tensor]
  type = ComputeElasticityTensor
  C_ijkl = '6 7 8 9 10'
  fill_method = axisymmetric_rz
  base_name = alpha_phase
[../]
```

as in, I have two completely different elasticity tensors that I do some operations on in the material file. And just so I get you, is that the proper way to hardcode the custom name? As in, whatever I type for base_name + "_" + "elasticity_tensor"?

**Reply by GiudGiud** [reply]
Yes this seems fine. 
The proper way would be to have two parameters:
base_name_phase_1
base_name_phase_2 
retrieve them in two strings, and use that to retrieve the tensors. 
This will let you change the base name in `alpha_elasticity_tensor` in the future for example

**Reply by richmondodufisan** [reply]
I see. How then would I differentiate between them, though? Say I have:

```
_base_name_phase_1(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
_base_name_phase_2(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
```

how do I tell which block each base name is gotten from? How do I know _base_name_phase_1 is gotten from the "alpha_elasticity_tensor" block and _base_name_phase_2 is gotten from the "beta_elasticity_tensor" block?

**Reply by GiudGiud** [reply]
_base_name_phase_1(isParamValid("base_name_phase_1") ? getParam<std::string>("base_name_phase_1") + "_" : ""),
_base_name_phase_2(isParamValid("base_name_phase_2") ? getParam<std::string>("base_name_phase_2") + "_" : ""),

use two params

**Reply by richmondodufisan** [reply]
I see, and I do that in the input file as well? As in,

```
[./beta_elasticity_tensor]
  type = ComputeElasticityTensor
  fill_method = symmetric9
  C_ijkl = '1111 1122 1133 2222 2233 3333 2323 1313 1212'
  base_name_phase_2 = beta_phase
[../]
```

I didn't know I could add/modify the name of a parameter in a block apart from the predefined ones

**Reply by GiudGiud** [reply]
no in that one you leave base_name

**Reply by GiudGiud** [reply]
it s in the material where you use both that you have to have two parameters

**Reply by richmondodufisan** [reply]
what I don't understand then is how I will know which one is which. If I have

```
base_name_phase_1(isParamValid("base_name_phase_1") ? getParamstd::string("base_name_phase_1") + "" : ""),
base_name_phase_2(isParamValid("base_name_phase_2") ? getParamstd::string("base_name_phase_2") + "" : ""),
```

in the material file, how will I know if "base_name_phase_1" is being retrieved from the alpha or beta block in the input file? Where is "base_name_phase_1" defined in the input block, for me to know which one is which? I plan to store the value of the elasticity tensor gotten from base_name_phase_1 in some variable _C_alpha, and base_name_phase_2 in _C_beta. how do I know it's assigning them correctly?

**Reply by GiudGiud** [reply]
It will be in a parameter. in your custom material you will have:
```
[Materials]
  [mine]
     type = ....
     base_name_phase_1 = alpha_phase
      base_name_phase_2 = beta_phase
```

again, you can do this either with a parameter and be able to choose the names, or can hard code it

**Reply by richmondodufisan** [reply]
I see. Gotcha. Thank you so much for your patient explanations so far, I really appreciate it!

I have one more issue (with the ADMaterialConverter). So I made all my materials AD, and I have:

```
[convert_from_AD]
    type = MaterialADConverter
	ad_props_in = 'c_ijkl e_ij_0 mechanical_strain alpha_phase_elasticity_tensor beta_phase_elasticity_tensor e_sfts'
	reg_props_out = 'reg_c_ijkl reg_e_ij_0 reg_mechanical_strain reg_alpha_phase_elasticity_tensor reg_beta_phase_elasticity_tensor reg_e_sfts' 
[]
```

Each of the names here are names in quotes of the declared properties. for example in my material file I can retrieve this just fine,

`_user_SFTS(getADMaterialPropertyByName<RankTwoTensor>("e_sfts")),`

which is gotten from a ADGenericConstantRankTwoTensor in my input file. In the same material file, I also declared:

```
	_cust_eigenstr(declareADProperty<RankTwoTensor>("e_ij_0")),
	_elasticity_eta(declareADProperty<RankFourTensor>("c_ijkl"))
```

my understanding thus far is that the names in quotes are the ones passed between systems. Why can't my input file see those names then? For the ADMaterialConverter block, I have this error:

```
*** ERROR ***
Material has no property named: c_ijkl
```

And I get this error for each of the AD properties I try to convert to regular ones

**Reply by GiudGiud** [reply]
oh the ADMaterialConvertor is not equipped to deal with tensors. You need to use the tensor version. See in the source file:
```
registerMooseObject("MooseApp", MaterialADConverter);
registerMooseObject("MooseApp", RankFourTensorMaterialADConverter);
registerMooseObject("MooseApp", RankTwoTensorMaterialADConverter);
```

**Reply by richmondodufisan** [reply]
Ah, perfect. Thank  you!!
